app/src/main/java/org/sagebionetworks/research/mpower/ActivityRecorderService.kt

```diff
@@ -0,0 +1,280 @@
+                }
+    }
+
+    private fun taskResultObserver(taskResult: TaskResult?) {
+        Log.d(TAG, "Observed TaskResult: {$taskResult}")
+    }
+
+    override fun onStartCommand(intent: Intent?, flags: Int, startId: Int): Int {
+        Log.d(TAG, "onStartCommand: startId = $startId, isRecording = $isRecording")
+
+        intent?.let {
+            when (val event = it.getIntExtra(EVENT, -1)) {
+                STARTED_WALKING -> startRecording()
+                STOPPED_WALKING -> stopRecording()
+                else -> throw IllegalArgumentException("EVENT is invalid: $event")
+            }
+        }
+
+        return START_REDELIVER_INTENT
+    }
+
+    private fun timeToDateStr(dateInMs: Long): String {
+        return if (dateInMs > 0) {
+            SimpleDateFormat(TIME_PATTERN, Locale.US).format(Date(dateInMs))
+        } else {
+            ""
+        }
+    }
+
+    //WALK START EVENT - Ignore if already recording
+    // Call startForeground with notification to show user - need design from Design team
+    // Create Task
+    // Initialize RecorderManager
+    // Record 30 seconds of walk data
+    // Stop recording
+    // Save to Bridge
+    // Stop service
+    private fun startRecording() {
+        if (!isRecording) {
+            val sharedPrefs = getSharedPreferences(TRANSITION_PREFS, Context.MODE_PRIVATE)
+            val lastRecordedAt = sharedPrefs.getLong(LAST_RECORDED_AT, -1)
+            val now = Date().time
+
+            Log.d(TAG, "-- lastRecordedAt: ${timeToDateStr(lastRecordedAt)}, now: ${timeToDateStr(now)}")
+
+            if (now - lastRecordedAt > MIN_FREQUENCY_MS) {
+                Log.d(TAG, "-- startRecording ${timeToDateStr(now)}")
+                isRecording = true
+                startForeground()
+//                CoroutineScope(Dispatchers.Default).launch {
+//                    delay(MAX_RECORDING_TIME_MS)
+//                    stopRecording()
+//                }
+                // NOTE: Currently RecorderService within RecorderManager is not bound, need to fix
+                // don't care about navDirection
+                recorderManager.onStepTransition(null, startStep, 0)
+            } else {
+                Log.d(TAG, "-- NOT RECORDING - ONLY RECORD ONCE EVERY $MIN_FREQUENCY_MS")
+                stopSelf()
+            }
+        }
+    }
+
+    private fun startForeground() {
+        createNotificationChannel()
+        val notification: Notification = NotificationCompat.Builder(this, getString(R.string.foreground_channel_id))
+            .setContentTitle(getText(R.string.recording_notification_title))
+            .setContentText(getText(R.string.recording_notification_message))
+            .setSmallIcon(R.drawable.ic_launcher_foreground)
+            .build()
+        startForeground(FOREGROUND_NOTIFICATION_ID, notification)
+    }
+
+    private fun createNotificationChannel() {
+        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
```

Comment:
Does this check buy us anything?

Reply:
yea. android 10+ will require the use of channels for notifications. before 10 there are no concept of channels.

Reply:
Haha!  I misread it as "dot zero", not "dot oh".